core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2382,44 +2382,27 @@ public UfsInfo getUfsInfo(long mountId) {
     return new ArrayList<>(lostFiles);
   }
 
-  @Override
-  public long loadMetadata(AlluxioURI path, LoadMetadataContext context)
-      throws BlockInfoException, FileDoesNotExistException, InvalidPathException,
-      InvalidFileSizeException, FileAlreadyCompletedException, IOException, AccessControlException {
-    try (RpcContext rpcContext = createRpcContext();
-         LockedInodePath inodePath = mInodeTree.lockInodePath(path, LockPattern.READ);
-         FileSystemMasterAuditContext auditContext =
-             createAuditContext("loadMetadata", path, null, inodePath.getParentInodeOrNull())) {
-      if (context.getOptions().getCreateAncestors()) {
-        auditContext.setSrcInode(inodePath.getLastExistingInode());
-      }
-      try {
-        mPermissionChecker.checkParentPermission(Mode.Bits.WRITE, inodePath);
-      } catch (AccessControlException e) {
-        auditContext.setAllowed(false);
-        throw e;
-      }
-      loadMetadataInternal(rpcContext, inodePath, context);
-      // Re-traverse to pick up loaded inodes.
-      inodePath.traverse();
-      auditContext.setSrcInode(inodePath.getInode()).setSucceeded(true);
-      return inodePath.getInode().getId();
-    }
-  }
-
   /**
    * Loads metadata for the object identified by the given path from UFS into Alluxio.
    *
-   * If the object is loaded, the inode path will
+   * This operation requires users to have WRITE permission on the path
+   * and its parent path if path is a file, or WRITE permission on the
+   * parent path if path is a directory.
    *
    * @param rpcContext the rpc context
    * @param inodePath the path for which metadata should be loaded
    * @param context the load metadata context
+   * @throws AccessControlException if permission checking fails
+   * @throws BlockInfoException if an invalid block size is encountered
+   * @throws FileAlreadyCompletedException if the file is already completed
+   * @throws FileDoesNotExistException if there is no UFS path
+   * @throws InvalidFileSizeException if invalid file size is encountered
+   * @throws InvalidPathException if invalid path is encountered
    */
   private void loadMetadataInternal(RpcContext rpcContext, LockedInodePath inodePath,
       LoadMetadataContext context)
-      throws InvalidPathException, FileDoesNotExistException, BlockInfoException,
-      FileAlreadyCompletedException, InvalidFileSizeException, AccessControlException, IOException {
+      throws AccessControlException, BlockInfoException, FileAlreadyCompletedException,
+      FileDoesNotExistException, InvalidFileSizeException, InvalidPathException, IOException {
```

Comment:
Does it make sense to keep the internal method?

Reply:
What I found is that the `loadMetadataInternal` method is actually used in multiple places outside of being called from the Alluxio client. It is called other places depending on other RPCs' `LoadMetadataType` and `LoadMetadataInterval`.

i.e.

- `syncMetadataInternal` calls `loadMetadataInternal`
- `loadMetadataIfNotExist` calls `loadMetadataInternal`.
  - `getFileId` calls `loadMetadataIfNotExist`
  - `getFileInfo` calls `loadMetadataIfNotExist`
  - `listStatus` calls `loadMetadataIfNotExist`

I think the amount of usage warrants that we keep it.

Reply:
The drawback is what you mentioned above, having to keep `LoadMetadataPOptions`. Would we be able to refactor this such that the core logic can be reused but took inputs based on the other options from the appropriate callers?

Reply:
Well the original deprecation was simply the explicit `loadMetadata` method in the client. The current behavior to loading metadata from a UFS by setting `LoadMetadataType` and `LoadMetadataInterval` has not been deprecated/there hasn't been any discussion about changing the behavior. This is regarding when users utilize `ONCE`, `NEVER`, and `ALWAYS`. I think changes to that behavior would be out of scope for this PR.

Reply:
Currently we have a `*Context` and `*POptions` which maps 1-to-1 with RPCs. With this change, we no longer have the `loadMetadata` RPC but still have the `loadMetadataContext` and the associated `LoadMetadataPOptions` (this should be unnecessary since we do not need to send it over the wire even). My question is if we can remove the context and options by refactoring the code such that the callers of `LoadMetadataInternal` can incorporate the options into their own respective options or the common options.

Reply:
So looking again - I do like having 1-to-1 RPC to proto mappings. However, I guess what I'm seeing now is that if we remove the `LoadMetadataPOptions` - then we still need some object which is able to hold all the options that currently exist within it. It would basically just be a hand-written version of the current `LoadMetadataPOptions` - is there a point on using effort to write a class which basically already exists?

Reply:
Removing it would only make sense if we could merge the fields in the options into existing options. Let's consider that in the future, I've created: https://github.com/Alluxio/alluxio/issues/8591